realm/src/main/java/io/realm/RealmList.java

```diff
@@ -183,6 +191,10 @@ public E set(int location, E object) {
 
     // Transparently copies a standalone object or managed object from another Realm to the Realm backing this RealmList.
     private E copyToRealmIfNeeded(E object) {
+        if (!(realm instanceof Realm))  {
+            throw new IllegalStateException("Not supported for RealmList containing DynamicRealmObjects");
+        }
+        Realm realm = (Realm) this.realm;
```

Comment:
No, this is the semantics we are already using when doing normal `realm.copyToRealm()`, it doesn't modify the existing object, but creates a new copy that is inserted. So this method actually does create a copy.